Add query to detect out-of-order cpp/linux privilege dropping

theopolis opened the pull request:
Hi folks, I am working on this query (my first!) to detect usage of `setuid` before `setgid`, `initgroups`, etc and when error checking is missing. This situation indicates an attempt to drop privileges incorrectly, and the most important part of the bug is the lack of error handling. This would effect setuid/setgid-bit executables and daemons run as root with the intention to later dropping privileges. I added inline comments about my thoughts on false positives and false negatives.

I think this category of checks can be extended, for example another query can be added to find instances where `initgroups` or `setgroups` (dropping supplemental groups) are missing or where capabilities are not dropped, etc. For this reason I created a subfolder for the query, test code, and helpfile.

I tried to create the small set of test cases to improve accuracy. I also browsed GitHub for about 100 projects that may be impacted, by looking for usage of `setuid`. At this point I am mildly confident this is working as intended. But this is my first time using CodeQL and I am sure it could use a lot of feedback. On my mind are: is this performance enough, am I using the QL stdlib correctly, can I simplify the logic, is there a way to test it on a larger set of projects?

JLLeitschuh commented:
Very cool. I have a few scripts that can let you bulk import projects into your LGTM account so you can run this query against large numbers of projects.

I see you're not in the slack channel. I recommend reaching out to the GitHub Security Lab to get an invite:
https://twitter.com/ghsecuritylab

theopolis commented:
Wow, thank you so much for code review @rdmarsh2! It will take me some time to apply the feedback because I’d like to dig into some of what you’re mentioning.

theopolis commented:
Alright, I re-tested this against my sample set of projects (n=100). All of these projects use `setuid`-like functions in one way or another.

Here are some examples of the follow up
- Lightning-fast improvements from LXC https://github.com/lxc/lxc/pull/3265
- True-positive and fixed https://github.com/swaywm/sway/issues/4990
- True-positive and fixed https://github.com/DNS-OARC/dnscap/issues/208
- Created an issue, I will try to follow up (confirm impact/write PR) later: https://github.com/darkk/redsocks/issues/149
- Additional true-positives one dormant projects, will consider opening a PR later.

It is interesting if you take away the requirement for the result to flow into a condition. These results are more prone to false positives but are fun for code reviewing.

theopolis commented on CodeQL_Queries/cpp/PrivilegeDropping/PrivilegeDroppingOutoforder.ql:
I am going to pass on this for the moment. I do not think this happens often (where someone ignores the result for specific values of `errno`). My goal with the comment was to highlight that this additional checking should be considered.

If we are able to run the query on a massive set of projects and find that adding a check here is worthwhile I am happy to follow up and explore adding a guard. :) (I am going to go learn more about guard implementations right now.)

JLLeitschuh commented:
Should these reports have CVE numbers assigned to them?

theopolis commented:
> Should these reports have CVE numbers assigned to them?

IMO no unless there's precedent for documenting similar issues. But I can break the issues into specifics:
- [best practice] Target user's supplemental groups are not initialized (or alternatively source groups are not dropped). Meaning `initgroups` or `setgroups` are not called.
- [best practice] Order of dropping is correct but return codes are not checked. (for example [dimkr/loksh:misc.c#L298](https://github.com/dimkr/loksh/blob/db365e4fbbb2fd6d5bc5ec60d19771155f6f0542/misc.c#L298))
- [low priority weakness] Order is incorrect and return codes are not checked. This is what the query detects as written here.

The SwayVM case falls into the last category and I can see documenting that as a CVE. I will look for precedent this weekend and we can apply consistency.

nicowaisman commented:
Hey @theopolis, 
  A couple of logistic questions. The PR actually need to happen on https://github.com/Semmle/ql where CodeQL main queries are hosted (In this query case, under cpp section).
  Then we ask contributors, to [create an issue](https://github.com/github/securitylab/issues/new?assignees=&labels=All+For+One&template=all-for-one.md&title=%5BUSERNAME%5D%3A+%5BSUMMARY%5D)  in this repo to keep a track of each submission.

Let me know if you have any question

theopolis commented:
Hi @nicowaisman,

> The PR actually need to happen on https://github.com/Semmle/ql where CodeQL main queries are hosted (In this query case, under cpp section).

I am not confident enough that this query should be added to the the set of main queries. I'd rather keep it in the group of queries in this repo unless someone can run it across all of the LGTM projects and comment on the false positive rate.

> Then we ask contributors, to create an issue in this repo to keep a track of each submission.

I was mostly interested in learning CodeQL and getting feedback on my query (which I have, and the feedback has been great!) Since this query finds issues that are focused on correctness around mitigations it feels a little below the bar for qualifying for a bounty. If you think it does I'd rather pursue the [bug slayer template](https://securitylab.github.com/bounties) due to my concern above about confidence.

> 4. Open a pull request in the security-lab repo with a single CodeQL query. See the contribution guidelines for more details.
> 5. Create an issue using the bug slayer template. The issue should link to your pull request and contain a detailed report of the vulnerabilities your query finds. It should include a description of the vulnerabilities, their associated CVEs, and how the query allowed you to find them. Pull requests without an accompanying issue cannot be considered.

JLLeitschuh commented:
The way the program is setup, you can actually do both with the same vuln.

You can use the query to find vulns in open source projects and get CVEs assigned for it (you need a minimum of 4 for Bug Slayer). That's the first BB program.

Then for the second, submit it to the Semmle/ql repo for the All For One and that's a second BB submission.

I'm doing that here with a query I already had accepted to the 'All-For-One' program:

 1. All-For-One - Bountied - #21 
 2. Bug Hunter - Pending - #38

theopolis commented:
Ok so my next step is to open a similar PR for the `ql` repo. This may take several days.
